WalletWasabi/WabiSabi/Client/SmartRequestNode.cs

```diff
@@ -0,0 +1,170 @@
+
+			// TODO remove
+			// Add back any zero credentials that do have a corresponding edge
+			// (and therefore a task completion source), so that the
+			// `amountCredentials` enumerator matches the `amounts` enumerator.
+			// Assumes that `amounts` is ordered by value descending, so zero
+			// credentials are at the end. (same for vsizes).
+			amountCredentials = amountCredentials.Concat(Enumerable.Range(0, AmountCredentialTasks.Count() - amountCredentials.Count()).Select(_ => ZeroAmountCredentialPool.GetZeroCredential()));
+			vsizeCredentials = vsizeCredentials.Concat(Enumerable.Range(0, VsizeCredentialTasks.Count() - vsizeCredentials.Count()).Select(_ => ZeroVsizeCredentialPool.GetZeroCredential()));
+
+			foreach ((TaskCompletionSource<Credential> tcs, Credential credential) in AmountCredentialTasks.Zip(amountCredentials))
+			{
+				tcs.SetResult(credential);
+			}
+			foreach ((TaskCompletionSource<Credential> tcs, Credential credential) in VsizeCredentialTasks.Zip(vsizeCredentials))
+			{
+				tcs.SetResult(credential);
+			}
+		}
+
+		public async Task StartOutputRegistrationAsync(
+			BobClient bobClient,
+			Money effectiveCost,
+			Script scriptPubKey,
+			CancellationToken cancellationToken)
+		{
+			await Task.WhenAll(AmountCredentialToPresentTasks.Concat(VsizeCredentialToPresentTasks)).ConfigureAwait(false);
+			IEnumerable<Credential> inputAmountCredentials = AmountCredentialToPresentTasks.Select(x => x.Result);
+			IEnumerable<Credential> inputVsizeCredentials = VsizeCredentialToPresentTasks.Select(x => x.Result);
+
+			await bobClient.RegisterOutputAsync(
+				effectiveCost,
+				scriptPubKey,
+				inputAmountCredentials,
+				inputVsizeCredentials,
+				cancellationToken).ConfigureAwait(false);
+		}
+
+		private IEnumerable<long> AddExtraCredential(IEnumerable<long> valuesToRequest, IEnumerable<Credential> presentedCredentials)
+		{
+			var nonZeroValues = valuesToRequest.Where(v => v > 0);
+
+			if (nonZeroValues.Count() == ProtocolConstants.CredentialNumber)
+			{
+				return nonZeroValues;
+			}
+
+			var missing = presentedCredentials.Sum(cr => (long)cr.Amount.ToUlong()) - valuesToRequest.Sum();
+
+			if (missing > 0)
+			{
+				nonZeroValues = nonZeroValues.Append(missing);
+			}
+
+			var additionalZeros = ProtocolConstants.CredentialNumber - nonZeroValues.Count();
+
+			return nonZeroValues.Concat(Enumerable.Repeat(0L, additionalZeros));
+		}
+
+		private IEnumerable<long> AddExtraCredential(IEnumerable<long> valuesToRequest, long sum)
+		{
+			var nonZeroValues = valuesToRequest.Where(v => v > 0);
+
+			if (nonZeroValues.Count() == ProtocolConstants.CredentialNumber)
+			{
+				return nonZeroValues;
+			}
+
+			var missing = sum - valuesToRequest.Sum();
+
+			if (missing > 0)
+			{
+				nonZeroValues = nonZeroValues.Append(missing);
+			}
+
+			var additionalZeros = ProtocolConstants.CredentialNumber - nonZeroValues.Count();
+
+			return nonZeroValues.Concat(Enumerable.Repeat(0L, additionalZeros));
+		}
+	}
```

Comment:
Seems like this logic would be cleaner in `InputNode` and `ReissuanceNode`, perhaps as `CredentialAmountsToRequest` or something... can be added to #5944

Reply:
I guess it is possible, concept ACK. Follow-up PR.

Reply:
definitely in followup